soccer/planning/planner/CollectPlanner.cpp

```diff
@@ -0,0 +1,470 @@
+    if (!averageBallVelInitialized) {
+        averageBallVel = ball.velocity;
+        averageBallVelInitialized = true;
+    } else {
+        averageBallVel = *_targetPointAveragingGain * averageBallVel +
+                         (1 - *_targetPointAveragingGain) * ball.velocity;
+    }
+
+    // Approach direction is the direction we move towards the ball and through
+    // it
+    if (ball.velocity.mag() < *_ballSpeedApproachDirectionCutoff) {
+        // Move directly to the ball
+        approachDirection = (ball.position - startInstant.position()).norm();
+    } else {
+        // Approach the ball from behind
+        approachDirection = -averageBallVel.norm();
+    }
+
+    // Check if we should transition to control from approach
+    processStateTransition(ball, startInstant);
+
+    switch (currentState) {
+        // Moves from the current location to the slow point of approach
+        case CourseApproach:
+            previous = courseApproach(planRequest, startInstant, obstacles,
+                                      dynamicObstacles);
+            break;
+        // Moves from the slow point of approach to just before point of contact
+        case FineApproach:
+            previous = fineApproach(planRequest, startInstant, obstacles,
+                                    dynamicObstacles);
+            break;
+        // Move through the ball and stop
+        case Control:
+            previous = control(planRequest, partialStartInstant, partialPath,
+                               obstacles, dynamicObstacles);
+            break;
+        default:
+            previous = invalid(planRequest, obstacles, dynamicObstacles);
+            break;
+    }
+
+    return previous;
+}
+
+void CollectPlanner::checkSolutionValidity(BallState ball, RobotInstant start) {
+    bool nearBall = (ball.position - start.position()).mag() <
+                    *_distCutoffToApproach + *_distCutoffToControl;
+
+    // Check if we need to go back into approach
+    //
+    // See if we are not near the ball and both almost stopped
+    if (!nearBall && currentState == Control) {
+        currentState = CourseApproach;
+        approachDirectionCreated = false;
+        controlPathCreated = false;
+    }
+}
+
+void CollectPlanner::processStateTransition(BallState ball,
+                                            RobotInstant startInstant) {
+    // Do the transitions
+    double dist =
+        (startInstant.position() - ball.position).mag() - Robot_MouthRadius;
+    double speedDiff = (startInstant.linear_velocity() - averageBallVel).mag() -
+                       *_touchDeltaSpeed;
+
+    // If we are in range to the slow dist
+    if (dist < *_approachDistTarget + Robot_MouthRadius &&
+        currentState == CourseApproach) {
+        currentState = FineApproach;
+    }
+
+    // If we are close enough to the target point near the ball
+    // and almost the same speed we want, start slowing down
+    // TODO(Kyle): Check for ball sense?
```

Comment:
Unrelated to this PR, but do we have angle information in these planners now? The old iteration only had XY information

Reply:
Yeah, that's a big change to how these planners work. They now plan angles on the trajectory directly rather than returning an angle function.

Reply:
That was a big problem I had when trying to "match" the state machines inside the planner and in python. It was possible to have the ball behind us but close enough such that the planner thought we had the ball, but the python side knew we didn't. I don't think I ever found a good hack to make it completely consistent

Reply:
I'm going to add a TODO issue for this. It should be much more reasonable to incorporate once we move to the "behavior planner" approach to these planners, where (in this case) it will actually have control and visibility over the dribbler.

Reply:
#1518